source/opt/types.cpp

```diff
@@ -383,63 +384,72 @@ void SampledImage::GetExtraHashWords(
   image_type_->GetHashWords(words, seen);
 }
 
-Array::Array(Type* type, uint32_t length_id, uint32_t spec_id)
-    : Type(kArray),
-      element_type_(type),
-      length_id_(length_id),
-      length_spec_id_(spec_id),
-      length_constant_type_(nullptr),
-      length_constant_words_() {
+Array::Array(const Type* type, const Array::LengthInfo& length_info)
+    : Type(kArray), element_type_(type), length_info_(length_info) {
+  assert(type != nullptr);
   assert(!type->AsVoid());
-  assert(spec_id != 0u);
-}
-
-Array::Array(Type* type, uint32_t length_id, const Type* constant_type,
-             Operand::OperandData constant_words)
-    : Type(kArray),
-      element_type_(type),
-      length_id_(length_id),
-      length_spec_id_(0u),
-      length_constant_type_(constant_type),
-      length_constant_words_(constant_words) {
-  assert(!type->AsVoid());
-  assert(constant_type && constant_type->AsInteger());
+  assert(length_info.opcode == SpvOpSpecConstant ||
+         length_info.opcode == SpvOpSpecConstantOp ||
+         length_info.opcode == SpvOpConstant);
+  // A spec constant leaf value must have a spec id.
+  assert((length_info.opcode != SpvOpSpecConstant) ||
+         (length_info.words.size() == 1));
+  // A constant must have literal words.
+  assert((length_info.opcode != SpvOpConstant) ||
+         (length_info.words.size() > 0));
 }
 
 bool Array::IsSameImpl(const Type* that, IsSameCache* seen) const {
   const Array* at = that->AsArray();
   if (!at) return false;
-  bool is_same = element_type_->IsSameImpl(at->element_type_, seen) &&
-                 HasSameDecorations(that);
-  // If it is a specialized constant
-  if (length_spec_id_ != 0u) {
-    // ensure they have the same SpecId
-    is_same = is_same && length_spec_id_ == at->length_spec_id_;
-  } else {
-    // else, ensure they have the same length literal number.
-    is_same =
-        is_same &&
-        length_constant_type_->IsSameImpl(at->length_constant_type_, seen) &&
-        length_constant_words_ == at->length_constant_words_;
+  bool is_same = element_type_->IsSameImpl(at->element_type_, seen);
+  is_same = is_same && HasSameDecorations(that);
+
+  // A valid array size is a numeric constant.
+  // So the cases are:
+  //  OpConstant, or
+  //  OpSpecConstant, with a SpecId decoration, or
```

Comment:
Is just spec id enough? Related to earlier comments.

Reply:
Fixed.